### DISCUSSION TITLE
multi-node run on HPC slurm system (NERSC)

### DISCUSSION THREAD

**Discussion by SchroederSa** [svenreiche/Genesis-1.3-Version4#200] [Q&A] [answered]
Hi,

I was setting up genesis4 simulations on the HPC system NERSC.
It is already much faster than on my local computer, but I would be happy to gain some further speedup. I don't quite get multi-node runs working and see only poor to no scaling with ``task. '' Do you have an idea how to speed things up? 

I have only a minimal understanding of HPC systems and slurm, so my example script might be off:

That's my submission file:
```
#!/bin/bash
#SBATCH --qos=regular
#SBATCH --time=00:40:00
#SBATCH -N 1
#SBATCH --ntasks=4
#SBATCH --constraint=cpu
#SBATCH -A m558
#SBATCH -J genesis
#SBATCH --error job-%j.err
#SBATCH --output job-%j.out
#SBATCH --signal=SIGTERM@100

module load python 
conda activate optimas_w_genesis

mpirun -np 4 genesis4 inputs_ngrid_0.in
```

Many thanks for your help!
Best
-Sarah

**Comment by svenreiche** [comment, marked as answer]
You are using the slurm batch systems which can be configure differently.
So most of this is guessing here, based on the slurm system at our cluster.

1) one should use srun to evoke the execution, which is a wrapper to the underlying HPC protocol (in this case it should be mpirun). srun also provides automatically the number of processes.

2) The nersc computer differentiates between nodes, tasks per nodes and cpu per tasks.
The corresponding arguments are 
--nodes=4
--ntasks-per-node=36
--cpu-per-tasks=1
This will spawn 144 processes for the MPI protocol.

I checked on the website that NERSC is doing it in a similar fashion as our cluster (see e.g. https://docs.nersc.gov/jobs/examples/ and there basic MPI job)

3) For best performance one should not use more than one CPU per task and disable hyperthreads. This is important if one wants to use OpenMP or CUDA, which is currently not supported by Genesis.



4) Here is the submission script for our cluster:
some comments we have to specify if we use the CPUs or the GPUs therefore we have to define the parameter -constraint
Also we have different queues selected by -partition keyword. I guess it is the same as the -qos at NERSC.

Best
Sven

#!/bin/bash -l
#SBATCH --job-name="genesis"
#SBATCH --account="psi"
#SBATCH --mail-type=ALL
#SBATCH --mail-user=sven.reiche@psi.ch
#SBATCH --time=00:10:00
#SBATCH --nodes=1
#SBATCH --ntasks-per-core=1
#SBATCH --ntasks-per-node=36
#SBATCH --cpus-per-task=1
#SBATCH --partition=debug
#SBATCH --constraint=mc
#SBATCH --hint=nomultithread

export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK

module load daint-mc
module load cray-hdf5-parallel
module load cray-fftw

srun genesis4 Aramis.in

**Comment by SchroederSa** [comment]
Many thanks!

Just report for others:
I actually had a package binding issue. When I set up the batch script as suggested, it started 36 times the same simulation instead of properly binding the cores. Sticking to mpirun -n is overcoming the problem for now. I may have to load the mpich package at the cluster before installing genesis in the conda environment such that srun is properly bound or building genesis from source.

Sarah
